File: Graphs/LC_417_Pacific_Atlantic_Water_Flow.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.18

class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        m, n = len(heights), len(heights[0])
        pac_seen, atl_seen = set(), set()

        def dfs(row, col, prev, seen):
            if row < 0 or row == m or col < 0 or col == n or (row, col) in seen or heights[row][col] < prev:
                return
            seen.add((row, col))
            dfs(row+1, col, heights[row][col], seen)
            dfs(row-1, col, heights[row][col], seen)
            dfs(row, col+1, heights[row][col], seen)
            dfs(row, col-1, heights[row][col], seen)
            return 
        for row in range(m):
            dfs(row, 0, -1, pac_seen)
            dfs(row, n-1, -1, atl_seen)
        
        for col in range(n):
            dfs(0, col, -1, pac_seen)
            dfs(m-1, col, -1, atl_seen)
        
        ans = []
        for row in range(m):
            for col in range(n):
                if (row, col) in pac_seen and (row, col) in atl_seen:
                    ans.append([row, col])
        return ans

File: Graphs/LC_417_Pacific_Atlantic_Water_Flow.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.18

class Solution {
public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        m = heights.size(), n = heights[0].size();
        vector<vector<int>> atl_seen(m, vector<int>(n, 0));
        vector<vector<int>> pac_seen(m, vector<int>(n, 0));
        // begin DFS from the side to collect all reachable nodes
        for (int row = 0; row < m; row++){
            dfs(heights, row, 0, pac_seen, -1);
            dfs(heights, row, n-1, atl_seen, -1);
        }
        for (int col = 0; col < n; col++){
            dfs(heights, 0, col, pac_seen, -1);
            dfs(heights, m-1, col, atl_seen, -1);
        }
        vector<vector<int>> ans;
        for (int row = 0; row < m; row++){
            for (int col = 0; col < n; col++){
                if (atl_seen[row][col] == 1 && pac_seen[row][col] == 1){
                    ans.push_back({row, col});
                }
            }
        }
        return ans;
    }
private:
    int m, n;

    void dfs(vector<vector<int>>& heights, int row, int col, vector<vector<int>>& seen, int prev){
        if (row < 0 || row == m || col < 0 || col == n || seen[row][col] == 1 || heights[row][col] < prev){
            return;
        }
        seen[row][col] = 1;
        dfs(heights, row+1, col, seen, heights[row][col]);
        dfs(heights, row-1, col, seen, heights[row][col]);
        dfs(heights, row, col+1, seen, heights[row][col]);
        dfs(heights, row, col-1, seen, heights[row][col]);
    }
};